### PR TITLE
Logger setup

### PR DESCRIPTION
**Pull request by Nexengineer**

### CODE REVIEW

File: src/In.ProjectEKA.HipService/Logger/Log.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
+
+namespace In.ProjectEKA.HipService.Logger
+{
+    using System;
+    using Serilog;
+    
+    public class Log
+    {
+        public static ILogger Logger
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/Logger/Log.cs]: Codacy Issue found: [Provide a getter for 'Logger' or replace the property with a 'SetLogger' method.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4954915)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 5
           

Complexity increasing per file
==============================
- src/In.ProjectEKA.OtpService/Otp/Logger/Log.cs  1
- src/In.ProjectEKA.HipService/Logger/Log.cs  1
         

Clones added
============
- src/In.ProjectEKA.HipService/Link/LinkPatientRepository.cs  2
- src/In.ProjectEKA.HipService/Program.cs  1
- src/In.ProjectEKA.OtpService/Otp/Logger/Log.cs  1
- src/In.ProjectEKA.OtpService/Program.cs  1
- src/In.ProjectEKA.HipService/Logger/Log.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4954915&bid=16375076)

File: src/In.ProjectEKA.OtpService/Otp/Logger/Log.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
+
+namespace In.ProjectEKA.OtpService.Otp.Logger
+{
+    using System;
+    using Serilog;
+    
+    public class Log
```
Comment (ganesanarun) [inline src/In.ProjectEKA.OtpService/Otp/Logger/Log.cs]: Codacy Issue found: [Add a 'protected' constructor or the 'static' keyword to the class declaration.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4954915)

File: src/In.ProjectEKA.HipService/Logger/Log.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
+
+namespace In.ProjectEKA.HipService.Logger
+{
+    using System;
+    using Serilog;
+    
+    public class Log
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/Logger/Log.cs]: Codacy Issue found: [Add a 'protected' constructor or the 'static' keyword to the class declaration.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4954915)

File: src/In.ProjectEKA.OtpService/Otp/Logger/Log.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    {
+        public static ILogger Logger
```
Comment (ganesanarun) [inline src/In.ProjectEKA.OtpService/Otp/Logger/Log.cs]: Codacy Issue found: [Provide a getter for 'Logger' or replace the property with a 'SetLogger' method.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4954915)

File: src/In.ProjectEKA.OtpService/Otp/Logger/LogTemplate.cs [hunk 5]
```diff
@@ -0,0 +1,7 @@
+namespace In.ProjectEKA.OtpService.Otp.Logger
+{
+    public class LogTemplate
```
Comment (ganesanarun) [inline src/In.ProjectEKA.OtpService/Otp/Logger/LogTemplate.cs]: Codacy Issue found: [Add a 'protected' constructor or the 'static' keyword to the class declaration.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4954915)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 5
           

Complexity increasing per file
==============================
- src/In.ProjectEKA.HipService/Logger/Log.cs  1
- src/In.ProjectEKA.OtpService/Otp/Logger/Log.cs  1
         

Clones added
============
- src/In.ProjectEKA.HipService/Logger/Log.cs  1
- src/In.ProjectEKA.OtpService/Program.cs  1
- src/In.ProjectEKA.HipService/Link/LinkPatientRepository.cs  2
- src/In.ProjectEKA.HipService/Program.cs  1
- src/In.ProjectEKA.OtpService/Otp/Logger/Log.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4954915&bid=16375076)

File: src/In.ProjectEKA.OtpService/Otp/Logger/Log.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
+
+namespace In.ProjectEKA.OtpService.Otp.Logger
+{
+    using System;
+    using Serilog;
+    
+    public static class Log
+    {
+        private static ILogger Logger
```
Comment (ganesanarun) [inline src/In.ProjectEKA.OtpService/Otp/Logger/Log.cs]: Codacy Issue found: [Provide a getter for 'Logger' or replace the property with a 'SetLogger' method.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4954915)

File: src/In.ProjectEKA.OtpService/Otp/Logger/LogTemplate.cs (hunk 5 above)
Comment (ganesanarun) [inline src/In.ProjectEKA.OtpService/Otp/Logger/LogTemplate.cs]: Codacy Issue found: [Add a 'protected' constructor or the 'static' keyword to the class declaration.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4954915)

File: src/In.ProjectEKA.OtpService/Otp/Logger/Log.cs (hunk 6 above)
Comment (ganesanarun) [inline src/In.ProjectEKA.OtpService/Otp/Logger/Log.cs]: Codacy Issue found: [Remove the unused private property 'Logger'.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4954915)

File: src/In.ProjectEKA.HipService/Logger/Log.cs [hunk 7]
```diff
@@ -0,0 +1,44 @@
+
+namespace In.ProjectEKA.HipService.Logger
+{
+    using System;
+    using Serilog;
+    
+    public static class Log
+    {
+        private static ILogger Logger
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/Logger/Log.cs]: Codacy Issue found: [Remove the unused private property 'Logger'.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4954915)

Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/Logger/Log.cs]: Codacy Issue found: [Provide a getter for 'Logger' or replace the property with a 'SetLogger' method.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4954915)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- src/In.ProjectEKA.OtpService/Otp/Logger/Log.cs  1
- src/In.ProjectEKA.HipService/Logger/Log.cs  1
         

Clones added
============
- src/In.ProjectEKA.HipService/Link/LinkPatientRepository.cs  2
- src/In.ProjectEKA.OtpService/Otp/Logger/Log.cs  1
- src/In.ProjectEKA.OtpService/Program.cs  1
- src/In.ProjectEKA.HipService/Program.cs  1
- src/In.ProjectEKA.HipService/Logger/Log.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4954915&bid=16375076)

File: src/In.ProjectEKA.HipService/In.ProjectEKA.HipService.csproj [hunk 8]
```diff
@@ -18,6 +18,15 @@
...
         <PackageReference Include="Optional" Version="4.0.0" />
+        <PackageReference Include="Serilog" Version="2.9.0" />
+        <PackageReference Include="Serilog.AspNetCore" Version="3.2.0" />
+        <PackageReference Include="Serilog.Enrichers.Environment" Version="2.1.3" />
+        <PackageReference Include="Serilog.Enrichers.Process" Version="2.0.1" />
+        <PackageReference Include="Serilog.Enrichers.Thread" Version="3.2.0-dev-00747" />
+        <PackageReference Include="Serilog.Settings.AppSettings" Version="2.2.2" />
+        <PackageReference Include="Serilog.Settings.Configuration" Version="3.1.1-dev-00216" />
+        <PackageReference Include="Serilog.Sinks.Console" Version="4.0.0-dev-00839" />
+        <PackageReference Include="Serilog.Sinks.ElasticSearch" Version="8.1.0-alpha0011" />
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/In.ProjectEKA.HipService.csproj]: Is it needed now? 
```suggestion
        <PackageReference Include="Serilog.Sinks.ElasticSearch" Version="8.1.0-alpha0011" />
```

Comment (ganesanarun) [reply src/In.ProjectEKA.HipService/In.ProjectEKA.HipService.csproj]: Same thing with other package, check whether all those are really necessary.

File: src/In.ProjectEKA.HipService/Logger/Log.cs [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+namespace In.ProjectEKA.HipService.Logger
+{
+    using System;
+    
+    public static class Log
+    {
+        public static void Information(string format, params object[] arg)
+        {
+            Serilog.Log.Information(format, arg);
+        }
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/Logger/Log.cs]: Can't the format be predefined as everything is custom here now?

File: src/In.ProjectEKA.HipService/Program.cs [hunk 10]
```diff
@@ -1,18 +1,29 @@
-﻿namespace In.ProjectEKA.HipService
+﻿
+
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/Program.cs]: Why there are spaces getting added?

File: src/In.ProjectEKA.HipService/Program.cs [hunk 11]
```diff
@@ -1,18 +1,29 @@
...
+    using Serilog;
     public class Program
     {
         public static void Main(string[] args)
         {
+            var configuration = new ConfigurationBuilder()
+                .AddJsonFile("appsettings.json")
+                .Build();
+            Log.Logger = new LoggerConfiguration()
+                .ReadFrom.Configuration(configuration)
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/Program.cs]:
```suggestion
               .ReadFrom.Configuration(host.Services.GetRequiredService<IConfiguration>())
```

File: src/In.ProjectEKA.HipService/Program.cs [hunk 12]
```diff
@@ -1,18 +1,29 @@
...
-
+    using Microsoft.Extensions.Configuration;
+    using Serilog;
     public class Program
     {
         public static void Main(string[] args)
         {
+            var configuration = new ConfigurationBuilder()
+                .AddJsonFile("appsettings.json")
+                .Build();
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/Program.cs]: Not needed.

File: src/In.ProjectEKA.HipService/Logger/Log.cs (hunk 9 above)
Comment (Nexengineer) [reply src/In.ProjectEKA.HipService/Logger/Log.cs]: format is defined in the template file

File: src/In.ProjectEKA.HipService/In.ProjectEKA.HipService.csproj (hunk 8 above)
Comment (ganesanarun) [reply src/In.ProjectEKA.HipService/In.ProjectEKA.HipService.csproj]:
```public static class Program
    {
        public static Task<int> Main(string[] args) => LogAndRunAsync(CreateHostBuilder(args).Build());

        public static async Task<int> LogAndRunAsync(IHost host)
        {
            if (host is null)
            {
                throw new ArgumentNullException(nameof(host));
            }

            // Use the W3C Trace Context format to propagate distributed trace identifiers.
            // See https://devblogs.microsoft.com/aspnet/improvements-in-net-core-3-0-for-troubleshooting-and-monitoring-distributed-apps/
            Activity.DefaultIdFormat = ActivityIdFormat.W3C;

            Log.Logger = CreateLogger(host);

            try
            {
                Log.Information("Started application");
                await host.RunAsync().ConfigureAwait(false);
                Log.Information("Stopped application");
                return 0;
            }
#pragma warning disable CA1031 // Do not catch general exception types
            catch (Exception exception)
#pragma warning restore CA1031 // Do not catch general exception types
            {
                Log.Fatal(exception, "Application terminated unexpectedly");
                return 1;
            }
            finally
            {
                Log.CloseAndFlush();
            }
        }

        public static IHostBuilder CreateHostBuilder(string[] args) =>
            new HostBuilder()
                .UseContentRoot(Directory.GetCurrentDirectory())
                .ConfigureHostConfiguration(
                    configurationBuilder => configurationBuilder
                        .AddEnvironmentVariables(prefix: "DOTNET_")
                        .AddIf(
                            args != null,
                            x => x.AddCommandLine(args)))
                .ConfigureAppConfiguration((hostingContext, config) =>
                    AddConfiguration(config, hostingContext.HostingEnvironment, args))
                .UseSerilog()
                .UseDefaultServiceProvider(
                    (context, options) =>
                    {
                        var isDevelopment = context.HostingEnvironment.IsDevelopment();
                        options.ValidateScopes = isDevelopment;
                        options.ValidateOnBuild = isDevelopment;
                    })
                .ConfigureWebHost(ConfigureWebHostBuilder)
                .UseConsoleLifetime();

        private static void ConfigureWebHostBuilder(IWebHostBuilder webHostBuilder) =>
            webHostBuilder
                .UseKestrel((builderContext, options) => options.AddServerHeader = false)
#if Azure
                .UseAzureAppServices()
#endif
                // Used for IIS and IIS Express for in-process hosting. Use UseIISIntegration for out-of-process hosting.
                .UseIIS()
                .UseStartup<Startup>();

        private static IConfigurationBuilder AddConfiguration(
            IConfigurationBuilder configurationBuilder,
            IHostEnvironment hostEnvironment,
            string[] args) =>
            configurationBuilder
                // Add configuration from the appsettings.json file.
                .AddJsonFile("appsettings.json", optional: true, reloadOnChange: false)
                // Add configuration from an optional appsettings.development.json, appsettings.staging.json or
                // appsettings.production.json file, depending on the environment. These settings override the ones in
                // the appsettings.json file.
                .AddJsonFile($"appsettings.{hostEnvironment.EnvironmentName}.json", optional: true, reloadOnChange: false)
                // Add configuration from files in the specified directory. The name of the file is the key and the
                // contents the value.
                .AddKeyPerFile(Path.Combine(Directory.GetCurrentDirectory(), "configuration"), optional: true)
                // This reads the configuration keys from the secret store. This allows you to store connection strings
                // and other sensitive settings, so you don't have to check them into your source control provider.
                // Only use this in Development, it is not intended for Production use. See
                // http://docs.asp.net/en/latest/security/app-secrets.html
                .AddIf(
                    hostEnvironment.IsDevelopment() && !string.IsNullOrEmpty(hostEnvironment.ApplicationName),
                    x => x.AddUserSecrets(Assembly.GetExecutingAssembly(), optional: true))
                // Add configuration specific to the Development, Staging or Production environments. This config can
                // be stored on the machine being deployed to or if you are using Azure, in the cloud. These settings
                // override the ones in all of the above config files. See
                // http://docs.asp.net/en/latest/security/app-secrets.html
                .AddEnvironmentVariables()
#if ApplicationInsights
                // Push telemetry data through the Azure Application Insights pipeline faster in the development and
                // staging environments, allowing you to view results immediately.
                .AddApplicationInsightsSettings(developerMode: !hostEnvironment.IsProduction())
#endif
                // Add command line options. These take the highest priority.
                .AddIf(
                    args != null,
                    x => x.AddCommandLine(args));

        private static Logger CreateLogger(IHost host) =>
            new LoggerConfiguration()
                .ReadFrom.Configuration(host.Services.GetRequiredService<IConfiguration>())
                .Enrich.WithProperty("Application", GetAssemblyProductName())
                .CreateLogger();

        private static string GetAssemblyProductName() =>
            Assembly.GetExecutingAssembly().GetCustomAttribute<AssemblyProductAttribute>().Product;
    }

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- src/In.ProjectEKA.HipService/Program.cs  1
- src/In.ProjectEKA.OtpService/Otp/Logger/Log.cs  1
- src/In.ProjectEKA.HipService/Logger/Log.cs  1
- src/In.ProjectEKA.OtpService/Program.cs  1
         

Clones added
============
- src/In.ProjectEKA.HipService/Program.cs  1
- src/In.ProjectEKA.OtpService/Otp/Logger/Log.cs  1
- src/In.ProjectEKA.HipService/Link/LinkPatientRepository.cs  2
- src/In.ProjectEKA.HipService/Logger/Log.cs  1
- src/In.ProjectEKA.OtpService/Program.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4954915&bid=16375076)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- src/In.ProjectEKA.HipService/Logger/Log.cs  1
- src/In.ProjectEKA.OtpService/Program.cs  1
- src/In.ProjectEKA.OtpService/Otp/Logger/Log.cs  1
- src/In.ProjectEKA.HipService/Program.cs  1
         

Clones added
============
- src/In.ProjectEKA.HipService/Logger/Log.cs  1
- src/In.ProjectEKA.OtpService/Program.cs  1
- src/In.ProjectEKA.HipService/Link/LinkPatientRepository.cs  2
- src/In.ProjectEKA.OtpService/Otp/Logger/Log.cs  1
- src/In.ProjectEKA.HipService/Program.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4954915&bid=16375076)

File: src/In.ProjectEKA.HipService/Program.cs [hunk 13]
```diff
@@ -1,18 +1,49 @@
...
+            finally
+            {
+                Log.CloseAndFlush();
+            }
         }
-
+        private static ILogger CreateLogger(IWebHost host) =>
+            new LoggerConfiguration()
+                .ReadFrom.Configuration(host.Services.GetRequiredService<IConfiguration>())
+                .CreateLogger();
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/Program.cs]:
```suggestion
                .CreateLogger();
                
```

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- src/In.ProjectEKA.HipService/Program.cs  1
- src/In.ProjectEKA.OtpService/Otp/Logger/Log.cs  1
- src/In.ProjectEKA.OtpService/Program.cs  1
- src/In.ProjectEKA.HipService/Logger/Log.cs  1
         

Clones added
============
- src/In.ProjectEKA.HipService/Link/LinkPatientRepository.cs  2
- src/In.ProjectEKA.HipService/Program.cs  1
- src/In.ProjectEKA.OtpService/Otp/Logger/Log.cs  1
- src/In.ProjectEKA.OtpService/Program.cs  1
- src/In.ProjectEKA.HipService/Logger/Log.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4954915&bid=16375076)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- src/In.ProjectEKA.HipService/Logger/Log.cs  1
- src/In.ProjectEKA.OtpService/Otp/Logger/Log.cs  1
- src/In.ProjectEKA.HipService/Program.cs  1
- src/In.ProjectEKA.OtpService/Program.cs  1
         

Clones added
============
- src/In.ProjectEKA.HipService/Logger/Log.cs  1
- src/In.ProjectEKA.OtpService/Otp/Logger/Log.cs  1
- src/In.ProjectEKA.HipService/Link/LinkPatientRepository.cs  2
- src/In.ProjectEKA.HipService/Program.cs  1
- src/In.ProjectEKA.OtpService/Program.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4954915&bid=16375076)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- src/In.ProjectEKA.HipService/Logger/Log.cs  1
- src/In.ProjectEKA.OtpService/Program.cs  1
- src/In.ProjectEKA.HipService/Program.cs  1
- src/In.ProjectEKA.OtpService/Otp/Logger/Log.cs  1
         

Clones added
============
- src/In.ProjectEKA.HipService/Logger/Log.cs  1
- src/In.ProjectEKA.OtpService/Program.cs  1
- src/In.ProjectEKA.HipService/Program.cs  1
- src/In.ProjectEKA.OtpService/Otp/Logger/Log.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4954915&bid=16375076)

**Pull request merged**